File: entities/message.py
from dataclasses import dataclass, field
import uuid

@dataclass
class Message:
    """
    A data structure representing a single email message.
    
    Fields:
    - message_id: Unique ID from email service
    - thread_id: ID of parent thread
    - sender: Email address of sender
    - recipients: List of recipient email addresses
    - subject: Email subject line
    - body: Main message content
    - date: When message was sent
    - labels: Any labels/categories applied to message
    """
    user_id: str 
    gmail_message_id: str
    header_message_id: str
    thread_id: str 
    labels: list = field(default_factory=list)
    snippet: str = ""
    history_id: str = ""
    sender: str = ""
    to: list = ""
    cc: list = ""
    reply_to_address: list = ""
    date: str = ""
    subject: str = ""
    body: str = ""
    summary: str = ""
    classification: str = ""
    in_reply_to: str = ""
    references: list = field(default_factory=list)
    response: str = ""
    thread_messages: list = field(default_factory=list)
    uuid: str = field(default_factory=lambda: str(uuid.uuid4()))
    # We use a lambda to generate a unique UUID if none is provided
    status: str = "unprocessed"

File: src/entities/message.ts
import { randomUUID } from 'node:crypto';

/** Result of classifying a message (labels are full, prefixed Gmail labels). */
export interface Classification {
  classification: string;
  priority: string;
  explanation: string;
}

/** Fields accepted when constructing a {@link Message}. */
export interface MessageOptions {
  userId: string;
  gmailMessageId: string;
  headerMessageId: string;
  threadId: string;
  labels?: string[];
  snippet?: string;
  historyId?: string;
  sender?: string;
  to?: string[];
  cc?: string[];
  replyToAddress?: string[];
  date?: string;
  subject?: string;
  body?: string;
  inReplyTo?: string;
  references?: string[];
}

/** A single email message tracked through the processing pipeline. */
export class Message {
  userId = '';
  gmailMessageId = '';
  headerMessageId = '';
  threadId = '';
  labels: string[] = [];
  snippet = '';
  historyId = '';
  sender = '';
  to: string[] = [];
  cc: string[] = [];
  replyToAddress: string[] = [];
  date = '';
  subject = '';
  body = '';
  inReplyTo = '';
  references: string[] = [];

  summary = '';
  classification: Classification | undefined = undefined;
  response = '';
  threadMessages: unknown[] = [];
  readonly uuid: string = randomUUID();
  status = 'unprocessed';

  constructor(options: MessageOptions) {
    // Copy only defined values so omitted optionals keep their class defaults.
    const defined = Object.fromEntries(
      Object.entries(options).filter(([, value]) => value !== undefined),
    );
    Object.assign(this, defined);
  }
}
